installer/ssh_cluster.go

```diff
@@ -0,0 +1,431 @@
+package installer
+
+import (
+	"crypto/md5"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/flynn/flynn/cli/config"
+	"github.com/flynn/flynn/pkg/httphelper"
+	"github.com/flynn/flynn/pkg/knownhosts"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/golang.org/x/crypto/ssh"
+	"github.com/flynn/flynn/Godeps/_workspace/src/golang.org/x/crypto/ssh/agent"
+)
+
+func (c *SSHCluster) Type() string {
+	return "ssh"
+}
+
+func (c *SSHCluster) Base() *BaseCluster {
+	return c.base
+}
+
+func (c *SSHCluster) SetBase(base *BaseCluster) {
+	c.base = base
+}
+
+func (c *SSHCluster) SetCreds(creds *Credential) error {
+	return nil
+}
+
+func (c *SSHCluster) SetDefaultsAndValidate() error {
+	c.ClusterID = c.base.ID
+
+	if err := c.validateInputs(); err != nil {
+		return err
+	}
+
+	c.base.SSHUsername = c.SSHLogin
+
+	if err := c.base.SetDefaultsAndValidate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *SSHCluster) validateInputs() error {
+	for _, ip := range c.TargetIPs {
+		if net.ParseIP(ip) == nil {
+			return httphelper.JSONError{
+				Code:    httphelper.ValidationErrorCode,
+				Message: fmt.Sprintf("%s is not a valid IP address", ip),
+			}
+		}
+	}
+	if c.SSHLogin == "" {
+		return httphelper.JSONError{
+			Code:    httphelper.ValidationErrorCode,
+			Message: "SSH login must be specified",
+		}
+	}
+	return nil
+}
+
+func (c *SSHCluster) Run() {
+	go func() {
+		defer c.base.handleDone()
+
+		steps := []func() error{
+			c.findSSHAuth,
+			c.saveInstanceIPs,
+			c.base.allocateDomain,
+			c.configureDNS,
+			c.installFlynn,
+			c.base.bootstrap,
+		}
+
+		for _, step := range steps {
+			if err := step(); err != nil {
+				if c.base.State != "deleting" {
+					c.base.setState("error")
+					c.base.SendError(err)
+				}
+				return
+			}
+		}
+
+		c.base.setState("running")
+
+		if err := c.base.configureCLI(); err != nil {
+			c.base.SendLog(fmt.Sprintf("WARNING: Failed to configure CLI: %s", err))
+		}
+	}()
+}
+
+func (c *SSHCluster) sshAgentAuth() ssh.AuthMethod {
+	if sock, err := net.Dial("unix", os.Getenv("SSH_AUTH_SOCK")); err == nil {
+		return ssh.PublicKeysCallback(agent.NewClient(sock).Signers)
+	}
+	return nil
+}
+
+type privateKeySigner struct {
+	Encrypted bool
+
+	pem       *pem.Block
+	key       *rsa.PrivateKey
+	publicKey ssh.PublicKey
+	path      string
+
+	base *BaseCluster
+}
+
+func (p privateKeySigner) Passphrase() string {
+	return p.base.PromptProtectedInput(fmt.Sprintf("Please enter the passphrase to unlock key: %s", p.path))
```

Comment:
We should cache the passphrase so that it can be re-used for the same key on a different host.